mantidproject/mantid
# Crystal Field single/multi spectra verification

Modifies the Python interface for the crystal field calculations to intepret single-element lists as scalars and to force using the single-spectrum fit function rather than the multi-spectra fit function. Allows the use of mixes of lists and scalars for the different input parameters. Also checks that the number of datasets implied by the different parameters, `Temperature`, `FWHM`, `IntensityScaling` and `InputWorkspace` are consistent.

**To test:**

The following script should not raise any errors:

```
#from CrystalField import CrystalFieldFit, CrystalField
import CrystalField
reload(CrystalField)
reload(CrystalField.function)
reload(CrystalField.fitting)

x = np.linspace(1,20)
ws1 = CreateWorkspace(x,x,x)
ws2 = CreateWorkspace(x,x*x,x)
ws3 = CreateWorkspace(x,x*x*x,x)
cf = CrystalField.fitting.CrystalField('Nd', 'C2v',B20=-0.4,  Temperature=[5,10], FWHM=[0.5,1,2])
cf.IntensityScaling = [1,2,3,4]
fit = CrystalField.fitting.CrystalFieldFit(Model=cf, InputWorkspace=[ws1],MaxIterations=200)
try:
    fit.fit()
except ValueError as e:
    print 'correctly raises error: %s' % (e)
# Bad fit string from IntensityScaling - longwinded error in C++ 

cf = CrystalField.fitting.CrystalField('Nd', 'C2v',B20=-0.4,  Temperature=[5,10], FWHM=[0.5,1,2])
cf.IntensityScaling = [1,2]
fit = CrystalField.fitting.CrystalFieldFit(Model=cf, InputWorkspace=[ws1],MaxIterations=200)
try:
    fit.fit()
except ValueError as e:
    print 'correctly raises error: %s' % (e)
# Bad fit string from FWHM - longwinded error in C++

cf = CrystalField.fitting.CrystalField('Nd', 'C2v',B20=-0.4,  Temperature=[5,10], FWHM=[0.5,1])
cf.IntensityScaling = [1,2]
fit = CrystalField.fitting.CrystalFieldFit(Model=cf, InputWorkspace=[ws1],MaxIterations=200)
try:
    fit.fit()
except ValueError as e:
    print 'correctly raises error: %s' % (e)
# Expected 2 input workspaces - cryptic error in C++

cf = CrystalField.fitting.CrystalField('Nd', 'C2v',B20=-0.4,  Temperature=[5], FWHM=0.5)
cf.IntensityScaling = [1]
fit = CrystalField.fitting.CrystalFieldFit(Model=cf, InputWorkspace=ws1,MaxIterations=1)
fit.fit()
# Bad fit string from IntensityScaling - longwinded error in C++, but should work.

cf = CrystalField.fitting.CrystalField('Nd', 'C2v',B20=-0.4,  Temperature=[5], FWHM=0.5)
cf.IntensityScaling = 1
fit = CrystalField.fitting.CrystalFieldFit(Model=cf, InputWorkspace=ws1,MaxIterations=1)
fit.fit()
# Expected scalar for Peaks object (part of CrystalField, defined by Temperature) - cryptic Python error
# But should work...

cf = CrystalField.fitting.CrystalField('Nd', 'C2v',B20=-0.4,  Temperature=[5], FWHM=0.5)
cf.IntensityScaling = 1
fit = CrystalField.fitting.CrystalFieldFit(Model=cf, InputWorkspace=[ws1],MaxIterations=1)
fit.fit()
```

Fixes #18465 .


*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
